Core/Src/pdu.c

```diff
@@ -15,26 +16,76 @@
 
 #define SHUTDOWN_ADDR PCA_I2C_ADDR_3
 #define CTRL_ADDR     PCA_I2C_ADDR_2
-#define RTDS_DURATION 2500
+#define RTDS_DURATION 1750 /* ms at 1kHz tick rate */
 
 static osMutexAttr_t pdu_mutex_attributes;
 
-static void rtds_shutoff_cb(void *pv_params)
+static uint8_t sound_rtds(pdu_t *pdu)
+{
+	if (!pdu)
+		return -1;
+
+	osStatus_t stat = osMutexAcquire(pdu->mutex, MUTEX_TIMEOUT);
+	if (stat)
+		return stat;
+
+	/* write RTDS over i2c */
+	HAL_StatusTypeDef error = pca9539_write_pin(
+		pdu->ctrl_expander, PCA_OUTPUT_1_REG, RTDS_CTRL, true);
+	if (error != HAL_OK) {
+		osMutexRelease(pdu->mutex);
+		return error;
+	}
+
+	osMutexRelease(pdu->mutex);
+
+	return 0;
+}
+
+static uint8_t rtds_shutoff(void *pv_params)
 {
 	pdu_t *pdu = (pdu_t *)pv_params;
 	osStatus_t stat = osMutexAcquire(pdu->mutex, MUTEX_TIMEOUT);
 	if (stat)
-		return;
+		return stat;
 
 	/* write RTDS over i2c */
 	HAL_StatusTypeDef error = pca9539_write_pin(
 		pdu->ctrl_expander, PCA_OUTPUT_1_REG, RTDS_CTRL, false);
 	if (error != HAL_OK) {
 		osMutexRelease(pdu->mutex);
-		return;
+		return error;
 	}
 
 	osMutexRelease(pdu->mutex);
+	return 0;
+}
+
+osThreadId_t rtds_thread;
+const osThreadAttr_t rtds_attributes = { .name = "RtdsThread",
+					 .stack_size = 300,
```

Comment:
im not sure if i missed this elsewhere, so defintely change if i did, but stack size allocations should always be powers of 2, no exceptions. (should make sense, because you cant fill up half of a byte in memory)

Reply:
I'm going to change to powers of 2, but based on what I'm reading from the docs, the stack_size attribute is the number of bytes allocated to the task, thus I do not think that it is necessary, and the case where you fill up half a byte of memory would not happen.